### PR TITLE
python: remove support for Python 2

### PR DESCRIPTION
**Pull request by trws**
This is a slightly bigger rework than I was planning after our last chat, but completely removes the requirement to use the make_build script and all the fragile regex munging we used to be doing in there.  (admittedly two sed commands remain, but they are much less prone to nastiness)

Main updates:
* build now creates preprocessed headers that are installed and could be used by a pypi package
* the build python scripts are explicit
* va_args functions should actually be bound now (not sure how useful this is)
* version is officially bumped from 2.7 to 3.6

There may be some dist problems left, but I think this covers what we were going for.

### CODE REVIEW

**Comment by trws** [comment]
Looks like the LGTM environment doesn't have the necessary python modules for python3.  @grondo, what's the best way to fix that up?

**Comment by grondo** [comment]
I don't know. I don't think we have control over the LGTM environment. Looks like it eventually commented?

Does this PR really add 100+ files under `src/bindings/python/pycparser-utils` or was that an accidental commit?

**Comment by SteVwonder** [comment]
Thanks @trws for putting this together!  I'm poking at it now.  Since we are vendoring the pycparser fake libc headers, do we also need to include pycparser's BSD-3 LICENSE file in that subdirectory?

**Comment by grondo** [comment]
> Since we are vendoring the pycparser fake libc headers, do we also need to include pycparser's BSD-3 LICENSE file in that subdirectory?

In my version of pycparser on Ubuntu the fake includes seem to be available in `/usr/share/python-pycparser/fake_libc_include`. Just curious why we're pulling them in instead?

File: .travis.yml [hunk 1]
```diff
@@ -28,8 +28,6 @@ jobs:
     - name: "Ubuntu: no configure flags"
       stage: test
       compiler: gcc
-      env:
-       - PYTHON_VERSION=2.7
     - name: "Ubuntu: py3.6 distcheck"
```
Comment (SteVwonder) [inline .travis.yml]: Maybe it is worth having some tests run explicitly against 3.6 and some against 3.7?  I'm not sure what the various distros have by default for python3.

File: configure.ac [hunk 2]
```diff
@@ -427,6 +433,10 @@ adl_RECURSIVE_EVAL([$luadir], fluxluadir)
 AS_VAR_SET(fluxluadir, $fluxluadir)
 AC_SUBST(fluxluadir)
 
+AS_VAR_SET(fluxbindingincludedir, $includedir/flux/_binding)
```
Comment (SteVwonder) [inline configure.ac]: To disambiguate from our other bindings, it might be worth doing the following:
```suggestion
AS_VAR_SET(fluxbindingincludedir, $includedir/flux/_py_binding)
```

File: src/bindings/python/_flux/_core_build.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ffi.set_source(
+    "_flux._core",
+    """
+#include <src/include/flux/core.h>
+
+
+void * unpack_long(ptrdiff_t num){
+  return (void*)num;
+}
+// TODO: remove this when we can use cffi 1.10
```
Comment (SteVwonder) [inline src/bindings/python/_flux/_core_build.py]: So close! TOSS3 has cffi 1.9.1 installed for python3.6 :sob:

**Comment by SteVwonder** [comment]
> In my version of pycparser on Ubuntu the fake includes seem to be available in /usr/share/python-pycparser/fake_libc_include. Just curious why we're pulling them in instead?

I believe `pycparser` would have to become another python dependency of flux-core, but if we are OK with doing that, then I believe you are correct that we can avoid pulling in all the various `fake_libc_include` files.

**Comment by SteVwonder** [comment]
> I believe pycparser would have to become another python dependency of flux-core

Scratch that; it looks like pycparser is already a dependency of cffi, which we require.  So maybe we can trivially depend on pycparser and it's utils being installed.

**Comment by trws** [comment]
It is already because of CFFI, but at least for a long time they didn’t distribute the headers with the pip package. If the packages now have them I’d be happy to rip them out.  Will do some checking on the current version pop package and dist packages.

File: src/bindings/python/_flux/_core_build.py (hunk 3 above)
Comment (trws) [reply src/bindings/python/_flux/_core_build.py]: I saw that... so, so close...

File: configure.ac (hunk 2 above)
Comment (trws) [reply configure.ac]: I'd be fine with that, but was wondering if it would make sense to leave it ambiguous since you could use the cleaned up headers with luajit ffi, for example, just as well as python cffi.  This is more indicative of how we use them, but there's no intrinsic python-ness involved in the headers.

File: .travis.yml (hunk 1 above)
Comment (trws) [reply .travis.yml]: Yeah, getting 3.8 in there and maybe 3.7 wouldn't be a bad idea.  We definitely have 3.7.2 in TCE on TOSS, so that would be a good one to keep tabs on, and since 3.8 has been out long enough we could add that one too.

**Comment by trws** [comment]
Ok, the dist packages (at least sometimes) include the fake headers, but a pip install does not install them.  They're only needed when building the preprocessed header from the flux headers, dealing with all the system header includes we have so we can use the actual preprocessor rather than my hacked together python version.    We could explore other options for this, including using the scripts in there to generate them as part of the build, still using my python script logic to strip system includes then using the compiler to preprocess after I guess, or wrapping system header includes with a macro, or... I dunno, there are lots of options but this seemed the least hacky one to try.

File: configure.ac (hunk 2 above)
Comment (SteVwonder) [reply configure.ac]: > was wondering if it would make sense to leave it ambiguous since you could use the cleaned up headers with luajit ffi, for example, just as well as python cffi. 

Ah! That makes a lot of sense.  Sounds good to leave it as is then.

**Comment by trws** [comment]
I'm going through trying to clean up the CI issues before making further changes here.  It looks like the LGTM issue is still persisting, but it's specific to the C/C++ side.  The python side works without issue and gave feedback,  but the C/C++ side can't find cffi and the configure fails.

**Comment by grondo** [comment]
Whoa. It looks like lgtm finds *and installs* dependencies _during_ the configure run:
```
[2020-03-09 21:48:45] [build] deptrace-server: File '/usr/bin/a2x' found in package asciidoc-base
[2020-03-09 21:48:45] [build] deptrace-server: Running installation program: /opt/deptrace/sudo-install
[2020-03-09 21:48:46] [build] checking for a2x... Reading package lists...
[2020-03-09 21:48:46] [build] Building dependency tree...
[2020-03-09 21:48:46] [build] Reading state information...
[2020-03-09 21:48:46] [build] The following additional packages will be installed:
[2020-03-09 21:48:46] [build]   asciidoc-common libxml2-utils
[2020-03-09 21:48:46] [build] Suggested packages:
[2020-03-09 21:48:46] [build]   asciidoc-doc docbook-utils source-highlight
[2020-03-09 21:48:46] [build] Recommended packages:
[2020-03-09 21:48:46] [build]   xmlto
[2020-03-09 21:48:46] [build] The following NEW packages will be installed:
[2020-03-09 21:48:46] [build]   asciidoc-base asciidoc-common libxml2-utils
```

Not sure why this isn't working for the python-cffi package, though. Maybe we have to force python3. It does look like lgtm has support for a .lgtm.yml:

https://help.semmle.com/lgtm-enterprise/user/help/lgtm.yml-configuration-file.html

**Comment by trws** [comment]
That's pretty amazing actually.  It looks like the C/C++ configure process of theirs isn't picking up the python package, while the python one does.  Oddly I don't see how to explicitly request some packages for C/C++, maybe we can just add in a command that explicitly installs python3-cffi from apt or something.

**Comment by grondo** [comment]
try `.lgtm.yml`:
```yaml
extraction:
  cpp:
    prepare:
      packages:
        - python3-cffi
```

**Comment by trws** [comment]
Thanks @grondo! I’ll try it as soon as I get the result back on 
whether my attempted automake fix flies.

**Comment by grondo** [comment]
Info on lgtm `deptrace`:
https://blog.semmle.com/cplusplus-projects-lgtm/

**Comment by grondo** [comment]
Here's the deptrace result for python-cffi for Python 2 (from our current builds)

Maybe the file that the python3-cffi check looks for isn't in the ubuntu db?

```
[2020-03-09 18:19:34] [build] deptrace-server: File '/usr/lib/python2.7/dist-packages/cffi/__init__.py' found in package python-cffi
[2020-03-09 18:19:34] [build] deptrace-server: Running installation program: /opt/deptrace/sudo-install
[2020-03-09 18:19:35] [build] checking for cffi.__version_info__ >= (1,1) in python module cffi... Reading package lists...
[2020-03-09 18:19:35] [build] Building dependency tree...
[2020-03-09 18:19:35] [build] Reading state information...
[2020-03-09 18:19:36] [build] The following additional packages will be installed:
[2020-03-09 18:19:36] [build]   python-cffi-backend python-ply python-pycparser
[2020-03-09 18:19:36] [build] Suggested packages:
[2020-03-09 18:19:36] [build]   python-ply-doc
[2020-03-09 18:19:36] [build] The following NEW packages will be installed:
[2020-03-09 18:19:36] [build]   python-cffi python-cffi-backend python-ply python-pycparser
```

**Comment by trws** [comment]
Maybe, that or somehow the detection isn't working with the python3 in there for some reason.  There's no deptrace output for the python3 stuff in the C/C++ build at all, which in and of itself seems a bit odd.

**Comment by trws** [comment]
Now that this is essentially passing (assuming the asan test comes back green), I could use some further feedback.  The main question is about handling the preprocessing.  This version vendors a kinda huge number of tiny generated files to deal with system includes.  I liked that as a thought because it uses the C preprocessor directly to do that whole job, but it's a lot of extra gack.  I think we have two main options here.

1. keep the headers
2. I rebuild a version of make_build.py that *only* does the include portion of its old job, leaving all the macro handling and so-forth to the C compiler

Preferences @SteVwonder, @grondo?

**Comment by garlick** [comment]
The vendored system headers displease me moderately FWIW.

**Comment by SteVwonder** [comment]
> keep the headers
> I rebuild a version of make_build.py that only does the include portion of its old job, leaving all the macro handling and so-forth to the C compiler
> Preferences @SteVwonder, @grondo?

>  I liked that as a thought because it uses the C preprocessor directly to do that whole job

I agree with you that having the preprocessor do all the work seems very nice and clean. The `make_build.py` has been surprisingly robust and worked well, but I typically get lost when I have to go spelunking into it (on the very rare occasion that is required). I think that is mainly because it is another layer of indirection when building the bindings and takes some time to grok each time, but anyways.... The number of vendored header files doen't bother me since they are tucked away in their own corner of the codebase, and they aren't that big size-wise so they shouldn't cause repo bloat. Maybe I'm not seeing the whole picture on their downside though. I do worry a bit about how fragile they are.  How frequently does libc change?

In summary, my preference is for keeping the headers, but I am not opposed to resurrecting the `make_build.py` script if that is where others are leaning.

**Comment by SteVwonder** [comment]
Looks like a hung test, just kicked it off again.  In the meantime, @trws, is this PR ready for another pass?  I'd be happy to take it for another spin if you are happy with it's current state.

**Comment by trws** [comment]
The Travis python versions have yet to be expanded on (and it could probably use a bit of rebase squishing), but otherwise I think this is good to go now

File: src/bindings/python/_flux/make_clean_header.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
+###############################################################
+# Copyright 2014 Lawrence Livermore National Security, LLC
```
Comment (SteVwonder) [inline src/bindings/python/_flux/make_clean_header.py]: Super minor nit, just b/c it's a new file.  Shouldn't block merging.

```suggestion
# Copyright 2020 Lawrence Livermore National Security, LLC
```

File: src/bindings/python/make_binding.py [hunk 5]
```diff
@@ -209,9 +209,12 @@ def process_header(f, including_path="."):
 {includes}
 
 ffi.cdef("""
+typedef int... ptrdiff_t;
```
Comment (SteVwonder) [inline src/bindings/python/make_binding.py]: Can this `make_binding.py` file be removed?  I'm not seeing any references to it anywhere anymore.

Comment (trws) [reply src/bindings/python/make_binding.py]: It can, I had been leaving it because I couldn't remember if there were any dependencies on it from outside of core.  I'll pull it out while I do my rebase this morning.

**Comment by trws** [comment]
This is rebased to clean up a lot of the churning and bring it up to date.  As of the last travis run, I seem to be having persistent trouble with one of the exec tests that has to do with successfully dealing with having /dev/null piped in, but I can't seem to repro this anywhere else.  @garlick, @grondo have you seen this anywhere else?

File: .travis.yml (hunk 1 above)
Comment (trws) [reply .travis.yml]: I'm building a new bionic docker image for this right now, we'll have 3.7 and 3.8 in travis shortly.

File: src/test/docker/bionic/Dockerfile [hunk 6]
```diff
@@ -5,10 +5,12 @@ LABEL maintainer="Tom Scogland <scogland1@llnl.gov>"
 # Update pkg caches, install latest pkg utils:
 RUN apt-get update \
  && apt-get -qq install -y --no-install-recommends \
-	apt-utils
+	apt-utils \
```
Comment (SteVwonder) [inline src/test/docker/bionic/Dockerfile]: Uber minor nit:

Looks like some tabs snuck in but the rest of the file uses space.

File: src/test/docker/bionic/Dockerfile [hunk 7]
```diff
@@ -23,10 +25,12 @@ RUN apt-get -qq install -y --no-install-recommends \
         lua5.2 \
         mpich \
         valgrind \
-	jq
+	jq \
```
Comment (SteVwonder) [inline src/test/docker/bionic/Dockerfile]: Same tabs vs spaces uber minor nit.

File: src/test/docker/bionic/Dockerfile (hunk 6 above)
Comment (trws) [reply src/test/docker/bionic/Dockerfile]: Good catch! Those may well have been there for quite a while, they are now gone.

**Comment by trws** [comment]
Thanks, although I think the credit for aspell goes to @grondo. =)

**Comment by grondo** [comment]
Nice!! :+1:

**Comment by trws** [comment]
> One thought on the travis python setup. I wonder if it's worthwhile to have one tester where there is explicitly both python 2 and python 3 installed to ensure that the configure picks up the py3 one (I tested that locally and it seems to work perfectly, but just to make sure there aren't any regressions).

Just realized I forgot to respond to this, most of the builders have both python2 and 3 installed, and several don't specify which to use, so we should be good there.

**Comment by trws** [comment]
The board is green, lgtm?

File: t/python/t0013-job-list-inactive.py [hunk 8]
```diff
@@ -238,4 +235,4 @@ def test_09_middle_timestamp_2(self):
     if rerun_under_flux(size=__flux_size(), personality="job"):
         from pycotap import TAPTestRunner
 
-        unittest.main(testRunner=TAPTestRunner())
+        unittest.main(testRunner=TAPTestRunner(), verbosity=2)
```
Comment (SteVwonder) [inline t/python/t0013-job-list-inactive.py]: Ugh. Forgot to hit `start a review` before approving.  Sorry.

Do we want to leave this verbosity up permanently?  If not, maybe we open an issue about tying to the `debug` env var much like `FLUX_PYCLI_LOGLEVEL` is in #2804 .

**Comment by grondo** [comment]
I reviewed quickly, so I could be wrong, but can `config/ax_python_devel.m4` be removed in this PR?

**Comment by SteVwonder** [comment]
> I reviewed quickly, so I could be wrong, but can config/ax_python_devel.m4 be removed in this PR?

I think it is still in use here: https://github.com/flux-framework/flux-core/pull/2805/files#diff-67e997bcfdac55191033d57a16d1408aR247

**Comment by grondo** [comment]
> I think it is still in use here: 

Oh, sorry! Only saw the removal. Thanks.

**Comment by trws** [comment]
Yup, that's right on all counts.  The ax_python_dev one sets up all the development paths and is used to drive the python path one in terms of which one actually has the right deps.  As it stands we need both for now.

**Comment by trws** [comment]
*sigh*
Much more going on here sadly.  So, it turns out the python test we found a race condition in wasn't being run, that's fixed.

A test much like one we already had NO_ASAN on in basic.t was hanging, added NO_ASAN to it.

Come to find out address sanitizer is broken on some certain versions of kernels from the last couple of years, bad relocations.  I apparently have one of these kernels on my dev box, yay.

Ended up writing a little chunk of shell script for sharness.sh that can impose a timeout on every single test from a global environment variable.  I'm planning to put that in a separate PR if this one passes the test runs this time, since it's how I managed to figure out which test was hanging on me (by setting it to a sane-ish value and seeing what failed).

**Comment by trws** [comment]
With this green, any chance of a MWP @SteVwonder, @grondo? I'll be putting up two other PRs in a bit for semi-related but distinct items.

**Comment by SteVwonder** [comment]
> With this green, any chance of a MWP @SteVwonder, @grondo?

I'm happy with this PR.  @garlick, @grondo: did you want to make a pass, or am I good to set the MWP label?

**Comment by chu11** [comment]
FYI, I had missed the conversation on slack regarding commit https://github.com/flux-framework/flux-core/pull/2805/commits/08b2806381b7a3c5a9a744fac6b74e63d552b8

I already had the fixes queued up in #2818, but the fixes in your patch are fine.

**Comment by grondo** [comment]
I haven't had a chance to do a good review, but I trust the other reviewers here anyway. I was going to give it a spin to see if anything came up, but been stuck in flux-security land. We can fix any minor issues with separate PRs. It will be nice to get this in IMO!

One thing: we should change the PR title or add a `Commit Message` section to the original post since the title we be used in the merge commit. :wink: (Maybe something like `python: remove support for Python 2`?)

**Comment by trws** [comment]
I'm sold on that, renamed the PR.

**Comment by grondo** [comment]
Excellent! I set merge-when-passing, will ask forgiveness please later!

**Comment by trws** [comment]
Huh, this is new... Travis says passing, GitHub apparently didn't get the memo.  Anyone know a way to poke it?

**Comment by grondo** [comment]
No, AFAIK, just have to be patient.

**Comment by trws** [comment]
In looking for solutions to this, most of what I'm finding are fixes for the old deprecated service api connector, which come to find out we're still using.  [post](https://travis-ci.community/t/known-issue-travis-ci-reports-expected-waiting-for-status-to-be-reported-on-the-github-status-api-but-the-status-never-arrives/1154)

Any issue with me setting it up to use the current github app integration?

**Comment by grondo** [comment]
No problem! Though I swear I had already done that long ago. I probably just thought I did!

**Comment by trws** [comment]
FWIW I would have sworn high and low that it was done quite a while ago 
too.  Not sure what’s going on but I’m going to hit the migrate 
button shortly, so if things go funny for a few minutes that’s 
what’s up.

**Comment by grondo** [comment]
This is slightly annoying. I'm kind of tempted to just remove branch protections for a moment and manually merge this one.

**Comment by trws** [comment]
Agreed, I'm trying to get the API bridge going again but something is definitely not behaving itself.

**Comment by grondo** [comment]
Since so many other PRs are dependent on this one, I went ahead and temporarily removed branch protections and hit the merge button. We'll figure out travis problems elsewhere.

**Comment by trws** [comment]
Thanks!  I got us transitioned to the app version, as soon as some of 
the builds start finishing things should work out.  One notable 
difference, we’re on travis-ci.com now, and the GitHub checks 
interface should start showing travis data.
